src/types.rs

```diff
@@ -65,20 +65,45 @@ impl From<Guid> for (u32, u32, u32, u32) {
 
 impl From<u128> for Guid {
     fn from(value: u128) -> Self {
-        let (a, b, c, d) = (
-            (value & 0xffffffff) as u32,
-            ((value >> 32) & 0xffffffff) as u32,
-            ((value >> 64) & 0xffffffff) as u32,
-            ((value >> 96) & 0xffffffff) as u32,
-        );
-        Guid::from((a, b, c, d))
+        Guid([
```

Comment:
is this really faster?

Reply:
The speedup is 4.6% on my machine in this synthetic benchmark. I would guess that the difference is use of two 64bit words vs four 32bit words, but I have not inspected the binaries to be sure.

```rust
// Original version: 26.103848ms
// Unrolled version: 24.937814ms
// Rolled version:   283.631294ms

use gvas::types::Guid;
use std::time::{Duration, Instant};

fn main() {
    // Generate some input data to test the function
    let input: u128 = 0x123456789abcdef0123456789abcdef;

    // Benchmark the original version
    let start = Instant::now();
    for _ in 0..1_000_000 {
        let _output = u128_to_bytes_original(input);
    }
    let elapsed_original = start.elapsed();

    // Benchmark the unrolled version
    let start = Instant::now();
    for _ in 0..1_000_000 {
        let _output = u128_to_bytes_unrolled(input);
    }
    let elapsed_unrolled = start.elapsed();

    // Benchmark the rolled version
    let start = Instant::now();
    for _ in 0..1_000_000 {
        let _output = u128_to_bytes_rolled(input);
    }
    let elapsed_rolled = start.elapsed();

    // Print the results
    println!("Original version: {:?}", elapsed_original);
    println!("Unrolled version: {:?}", elapsed_unrolled);
    println!("Rolled version:   {:?}", elapsed_rolled);
}

fn u128_to_bytes_original(input: u128) -> Guid {
    let (a, b, c, d) = (
        (input & 0xffffffff) as u32,
        ((input >> 32) & 0xffffffff) as u32,
        ((input >> 64) & 0xffffffff) as u32,
        ((input >> 96) & 0xffffffff) as u32,
    );
    Guid::from((a, b, c, d))
}

fn u128_to_bytes_rolled(input: u128) -> Guid {
    let mut bytes = [0; 16];
    for i in 0..16 {
        bytes[i] = ((input >> (8 * i)) & 0xff) as u8;
    }
    Guid(bytes)
}

fn u128_to_bytes_unrolled(input: u128) -> Guid {
    let mut bytes = [0; 16];
    bytes[0] = (input & 0xff) as u8;
    bytes[1] = ((input >> 8) & 0xff) as u8;
    bytes[2] = ((input >> 16) & 0xff) as u8;
    bytes[3] = ((input >> 24) & 0xff) as u8;
    bytes[4] = ((input >> 32) & 0xff) as u8;
    bytes[5] = ((input >> 40) & 0xff) as u8;
    bytes[6] = ((input >> 48) & 0xff) as u8;
    bytes[7] = ((input >> 56) & 0xff) as u8;
    bytes[8] = ((input >> 64) & 0xff) as u8;
    bytes[9] = ((input >> 72) & 0xff) as u8;
    bytes[10] = ((input >> 80) & 0xff) as u8;
    bytes[11] = ((input >> 88) & 0xff) as u8;
    bytes[12] = ((input >> 96) & 0xff) as u8;
    bytes[13] = ((input >> 104) & 0xff) as u8;
    bytes[14] = ((input >> 112) & 0xff) as u8;
    bytes[15] = ((input >> 120) & 0xff) as u8;
    Guid(bytes)
}
```

Reply:
I dug deeper. The improvement is coming from inlining the Guid::from call, but using u128 is slower than using four u32s. However, two inlined u64s is the fastest of these options:

```rust
// Inline64 version: 22.27163ms
// Inline32 version: 23.310772ms
// Unrolled version: 25.303099ms
// Original version: 26.121948ms
// Rolled version:   280.968913ms

use gvas::types::Guid;
use std::time::Instant;

fn main() {
    // Generate some input data to test the function
    let input: u128 = 0x123456789abcdef0123456789abcdef;

    // Benchmark the inlined 64bit version
    let start = Instant::now();
    for _ in 0..1_000_000 {
        let _output = u128_to_guid_inline_64(input);
    }
    println!("Inline64 version: {:?}", start.elapsed());

    // Benchmark the inlined 32bit version
    let start = Instant::now();
    for _ in 0..1_000_000 {
        let _output = u128_to_guid_inline_32(input);
    }
    println!("Inline32 version: {:?}", start.elapsed());

    // Benchmark the unrolled version
    let start = Instant::now();
    for _ in 0..1_000_000 {
        let _output = u128_to_guid_unrolled(input);
    }
    println!("Unrolled version: {:?}", start.elapsed());

    // Benchmark the original version
    let start = Instant::now();
    for _ in 0..1_000_000 {
        let _output = u128_to_guid_original(input);
    }
    println!("Original version: {:?}", start.elapsed());

    // Benchmark the rolled version
    let start = Instant::now();
    for _ in 0..1_000_000 {
        let _output = u128_to_guid_rolled(input);
    }
    println!("Rolled version:   {:?}", start.elapsed());
}

fn u128_to_guid_inline_64(input: u128) -> Guid {
    let (a, b) = (
        (input & 0xffffffffffffffff) as u64,
        ((input >> 64) & 0xffffffffffffffff) as u64,
    );
    Guid([
        (a & 0xff) as u8,
        ((a >> 8) & 0xff) as u8,
        ((a >> 16) & 0xff) as u8,
        ((a >> 24) & 0xff) as u8,
        ((a >> 32) & 0xff) as u8,
        ((a >> 40) & 0xff) as u8,
        ((a >> 48) & 0xff) as u8,
        ((a >> 56) & 0xff) as u8,
        (b & 0xff) as u8,
        ((b >> 8) & 0xff) as u8,
        ((b >> 16) & 0xff) as u8,
        ((b >> 24) & 0xff) as u8,
        ((b >> 32) & 0xff) as u8,
        ((b >> 40) & 0xff) as u8,
        ((b >> 48) & 0xff) as u8,
        ((b >> 56) & 0xff) as u8,
    ])
}

fn u128_to_guid_inline_32(input: u128) -> Guid {
    let (a, b, c, d) = (
        (input & 0xffffffff) as u32,
        ((input >> 32) & 0xffffffff) as u32,
        ((input >> 64) & 0xffffffff) as u32,
        ((input >> 96) & 0xffffffff) as u32,
    );
    Guid([
        (a & 0xff) as u8,
        ((a >> 8) & 0xff) as u8,
        ((a >> 16) & 0xff) as u8,
        ((a >> 24) & 0xff) as u8,
        (b & 0xff) as u8,
        ((b >> 8) & 0xff) as u8,
        ((b >> 16) & 0xff) as u8,
        ((b >> 24) & 0xff) as u8,
        (c & 0xff) as u8,
        ((c >> 8) & 0xff) as u8,
        ((c >> 16) & 0xff) as u8,
        ((c >> 24) & 0xff) as u8,
        (d & 0xff) as u8,
        ((d >> 8) & 0xff) as u8,
        ((d >> 16) & 0xff) as u8,
        ((d >> 24) & 0xff) as u8,
    ])
}

fn u128_to_guid_unrolled(input: u128) -> Guid {
    let mut bytes = [0; 16];
    bytes[0] = (input & 0xff) as u8;
    bytes[1] = ((input >> 8) & 0xff) as u8;
    bytes[2] = ((input >> 16) & 0xff) as u8;
    bytes[3] = ((input >> 24) & 0xff) as u8;
    bytes[4] = ((input >> 32) & 0xff) as u8;
    bytes[5] = ((input >> 40) & 0xff) as u8;
    bytes[6] = ((input >> 48) & 0xff) as u8;
    bytes[7] = ((input >> 56) & 0xff) as u8;
    bytes[8] = ((input >> 64) & 0xff) as u8;
    bytes[9] = ((input >> 72) & 0xff) as u8;
    bytes[10] = ((input >> 80) & 0xff) as u8;
    bytes[11] = ((input >> 88) & 0xff) as u8;
    bytes[12] = ((input >> 96) & 0xff) as u8;
    bytes[13] = ((input >> 104) & 0xff) as u8;
    bytes[14] = ((input >> 112) & 0xff) as u8;
    bytes[15] = ((input >> 120) & 0xff) as u8;
    Guid(bytes)
}

fn u128_to_guid_original(input: u128) -> Guid {
    let (a, b, c, d) = (
        (input & 0xffffffff) as u32,
        ((input >> 32) & 0xffffffff) as u32,
        ((input >> 64) & 0xffffffff) as u32,
        ((input >> 96) & 0xffffffff) as u32,
    );
    Guid::from((a, b, c, d))
}

fn u128_to_guid_rolled(input: u128) -> Guid {
    let mut bytes = [0; 16];
    for i in 0..16 {
        bytes[i] = ((input >> (8 * i)) & 0xff) as u8;
    }
    Guid(bytes)
}
```

Reply:
I also tested with a more realistic (randomized) workload and with `--release` enabled. It's a three way tie for first when adjusted for noise between runs, but all four of the inline algorithms are substantially faster than the original.

```
Original: 115.853329ms (4.67 x)
Inline64:  28.043403ms (1.13 x)
Inline32:  24.718602ms (1 x)
Unrolled:  24.683659ms (0.999 x)
Rolled:    24.401996ms (0.997 x)
```